addons/purchase/migrations/9.0.1.2/post-migration.py

```diff
@@ -194,30 +196,32 @@ def set_po_line_computed_rest(env):
     # be called in end-migration, where mrp module will be available for
     # sure in the environment if it is installed
     openupgrade.logged_query(
-        env.cr,
-        """
-            UPDATE purchase_order_line pol
-            SET qty_received = 0.0
-            FROM purchase_order po
-            WHERE
-                po.id = pol.order_id AND
-                po.state NOT IN ('purchase', 'done')
-        """
-    )
-    openupgrade.logged_query(
-        env.cr,
-        """
-            UPDATE purchase_order_line pol
-            SET qty_received = pol.product_qty
+        env.cr, """
+        UPDATE purchase_order_line pol
+        SET qty_received = sub.qty_received
+        FROM (
+            SELECT CASE
+                WHEN MIN(po.state) NOT IN ('purchase', 'done') THEN 0.0
+                WHEN MIN(po.state) IN ('purchase', 'done')
+                    AND MIN(pt.type) NOT IN ('consu', 'product')
+                    THEN MIN(pol2.product_qty)
+                ELSE SUM(CASE
+                    WHEN sm.state IS NULL OR sm.state != 'done' THEN 0.0
+                    ELSE COALESCE(sm.product_uom_qty / u.factor * u2.factor, 0)
```

Comment:
You're not checking that both UoMs belong to the same category, right? That's a behavioral change from: https://github.com/OCA/OpenUpgrade/blob/075cfa226b29819cae3ba1558fd8b3954c31aff0/addons/product/product.py#L122

Migration must fail in that case and provide a useful traceback. You can borrow some ideas from my code a little bit more above.

Reply:
Included check